### PR TITLE
dt1 editor: generate most of tile type images with canvas

### LINKED ISSUE #119: dt1 viewer: remove assets, generate with canvas
remove assets used in `hsassets/images`, generate equivalent images using a canvas.

### PR DESCRIPTION
**Pull request by gucio321**
hi there!
This resolves #119

### CODE REVIEW

File: hswidget/dt1widget/tile_type_image.go [hunk 1]
```diff
@@ -1,31 +1,93 @@
...
+			b := tiletypeimage.TileTypeImage(canvas, pos)
+			b.Floor().NorthDoor()
+		})
+	default:
+		return giu.ImageWithFile("./hsassets/images/" + tileImageFile)
+	}
+
+	// can't be reached
+	return nil
+}
```
Comment (gravestench) [inline hswidget/dt1widget/tile_type_image.go:85]: you can get away with writing half those lines with something like this:
```golang
func drawTileTypeImage(t d2enum.TileType) giu.Widget {
	tileImageFile := getTileTypeImage(t)

	canvas := giu.GetCanvas()
	pos := giu.GetCursorScreenPos()
	b := tiletypeimage.TileTypeImage(canvas, pos)

	fnMap := map[d2enum.TileType]func() {
		d2enum.TileFloor: func() { b.Floor() },
		d2enum.TileLeftWall: func() { b.Floor().NorthWall(true) },
		d2enum.TileRightWall: func() { b.Floor().NorthWall(true) },
		d2enum.TileRightPartOfNorthCornerWall: func() { b.Floor().WestWall(false).NorthWall(true) },
		d2enum.TileLeftPartOfNorthCornerWall: func() { b.Floor().WestWall(true).NorthWall(false) },
		d2enum.TileLeftEndWall: func() { b.Floor().EastWall() },
		d2enum.TileRightEndWall: func() { b.Floor().SoathWall() },
		d2enum.TileSouthCornerWall: func() { b.Floor().WestDoor() },
		d2enum.TileLeftWallWithDoor: func() { b.Floor().WestDoor() },
		d2enum.TileRightWallWithDoor: func() { b.Floor().NorthDoor() },
	}
	
	// default
	widget := giu.ImageWithFile("./hsassets/images/" + tileImageFile)
	
	if fnDraw, found := fnMap[t]; found {
		widget = giu.Custom(fnDraw)
	}
	
	return widget
}
```

**Review by gravestench — CHANGES_REQUESTED** [review]
Please change `drawTileTypeImage`, it's way too long.

```golang
func drawTileTypeImage(t d2enum.TileType) giu.Widget {
	canvas := giu.GetCanvas()
	pos := giu.GetCursorScreenPos()
	b := tiletypeimage.TileTypeImage(canvas, pos)

	fnMap := map[d2enum.TileType]func() {
		d2enum.TileFloor: func() { b.Floor() },
		d2enum.TileLeftWall: func() { b.Floor().NorthWall(true) },
		d2enum.TileRightWall: func() { b.Floor().NorthWall(true) },
		d2enum.TileRightPartOfNorthCornerWall: func() { b.Floor().WestWall(false).NorthWall(true) },
		d2enum.TileLeftPartOfNorthCornerWall: func() { b.Floor().WestWall(true).NorthWall(false) },
		d2enum.TileLeftEndWall: func() { b.Floor().EastWall() },
		d2enum.TileRightEndWall: func() { b.Floor().SoathWall() },
		d2enum.TileSouthCornerWall: func() { b.Floor().WestDoor() },
		d2enum.TileLeftWallWithDoor: func() { b.Floor().WestDoor() },
		d2enum.TileRightWallWithDoor: func() { b.Floor().NorthDoor() },
	}

	if fnDraw, found := fnMap[t]; found {
		return giu.Custom(fnDraw)
	}

	return giu.Dummy(0, 0)
}
```

**Comment by gucio321** [comment]
so, result looks like:

https://user-images.githubusercontent.com/73652197/121385751-abf9c300-c949-11eb-944f-5643b4493fd9.mp4

**Review by gravestench — CHANGES_REQUESTED** [review]
Remove `hswidget/dt1widget/tiletypeimage/trick.png`

**Comment by gucio321** [comment]
oh, sorry I forgot about it

File: hswidget/dt1widget/tile_type_image.go [hunk 2]
```diff
@@ -1,32 +1,33 @@
...
+			d2enum.TileRightEndWall:               func() { b.Floor().SoathWall() },
+			d2enum.TileSouthCornerWall:            func() { b.Floor().Corner() },
+			d2enum.TileLeftWallWithDoor:           func() { b.Floor().WestDoor() },
+			d2enum.TileRightWallWithDoor:          func() { b.Floor().NorthDoor() },
+		}
+
+		if creator, ok := lookup[t]; ok {
+			creator()
+		}
+	})
```
Comment (gravestench) [inline hswidget/dt1widget/tile_type_image.go]: What's the rationale here for deviating from what I suggested?

File: hswidget/dt1widget/tiletypeimage/tile_type_image.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	p2 := p3.Add(image.Pt(0, -wallH))
+
+	b.canvas.AddQuad(p1, p2, p3, p4, b.borderColor, thickness)
+	b.canvas.AddQuadFilled(p1, p2, p3, p4, b.wallColor)
+
+	return b
+}
+
+// SoathWall adds a wall on a soath
+func (b *Builder) SoathWall() *Builder {
```
Comment (gravestench) [inline hswidget/dt1widget/tiletypeimage/tile_type_image.go]: typo, should be `SouthWall`

File: hswidget/dt1widget/tile_type_image.go (hunk 2 above)
Comment (gucio321) [reply hswidget/dt1widget/tile_type_image.go]: you can call `drawTileTypeImage` and store a result in some variable, but it is important to call `GetCursorScreenPos` exactly at the time, when you're building it.

Comment (gucio321) [reply hswidget/dt1widget/tile_type_image.go]: other words, canvas should be drawn at position, when we put this `giu.Custom`, not where we call `drawTileTypeImage`

File: hswidget/dt1widget/tiletypeimage/tile_type_image.go (hunk 3 above)
Comment (gucio321) [reply hswidget/dt1widget/tiletypeimage/tile_type_image.go]: see eb8a792

File: go.mod [hunk 4]
```diff
@@ -27,4 +27,5 @@ require (
 	golang.org/x/mobile v0.0.0-20210527171505-7e972142eb43 // indirect
 	golang.org/x/net v0.0.0-20210525063256-abc453219eb5 // indirect
 	golang.org/x/sys v0.0.0-20210601080250-7ecdf8ef093b // indirect
+	golang.org/x/tools v0.1.2 // indirect
 )
```
Comment (gravestench) [inline go.mod]: why is go.mod and go.sum being modified in this PR?

**Review by gravestench — APPROVED** [review]

**Review by gravestench — DISMISSED** [review]

**Comment by gravestench** [comment]
please rebase off of upstream/master

File: go.mod (hunk 4 above)
Comment (gucio321) [reply go.mod]: good point

Comment (gucio321) [reply go.mod]: a4f2619

**Review by gravestench — APPROVED** [review]

**Pull request merged**